### PR TITLE
updated parent dependencies to align with 4.27

### PR DESCRIPTION
**Pull request by slaboure**
Fixed spotbug issue in collection

Updated dependencies with 4.27, and fixed a Findbug issue. Note that it might be interesting to think about updating https://github.com/whitesource/fs-agent as it is referencing old libraries and it forced a lot of dependency fixes in the Jenkins plugin

### CODE REVIEW

**Review by Wadeck — APPROVED** [review]
Seems good :) (not manually tested)

File: pom.xml [hunk 1]
```diff
@@ -77,22 +77,22 @@
 
     <properties>
         <!-- Minimal Jenkins version (2.14.0 ) required by for pipeline job -->
-        <jenkins.version>1.642.3</jenkins.version>
+        <jenkins.version>2.289.3</jenkins.version>
```
Comment (MarkEWaite) [inline pom.xml]: Usually we prefer the .1 release of the LTS.
```suggestion
        <jenkins.version>2.289.1</jenkins.version>
```

File: pom.xml [hunk 2]
```diff
@@ -159,7 +157,7 @@
         <dependency>
             <groupId>io.netty</groupId>
             <artifactId>netty-codec-http</artifactId>
-            <version>4.1.52.Final</version>
+            <version>5.0.0.Alpha2</version>
```
Comment (MarkEWaite) [inline pom.xml:159]: Need to revert this change.  Don't want an Alpha2 dependency

File: pom.xml [hunk 3]
```diff
@@ -174,7 +172,7 @@
         <dependency>
             <groupId>org.apache.httpcomponents</groupId>
             <artifactId>httpclient</artifactId>
```
Comment (MarkEWaite) [inline pom.xml]: Can use the httpcomponents API plugin available in Jenkins plugins and already installed by most users
```suggestion
            <groupId>org.jenkins-ci.plugins</groupId>
            <artifactId>apache-httpcomponents-client-4-api</artifactId>
```

File: pom.xml [hunk 4]
```diff
@@ -174,7 +172,7 @@
...
-            <version>4.5.3</version>
+            <version>4.5.13</version>
```
Comment (MarkEWaite) [inline pom.xml]: By using the Jenkins plugin and using the plugin bill of materials, don't need to specify a version.
```suggestion
```

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
The alpha dependency update needs to be reverted or pulled back to the most recent release.  Other changes can be considered optional

File: pom.xml [hunk 5]
```diff
@@ -77,22 +77,22 @@
...
-        <slf4j.version>1.7.5</slf4j.version>
+        <slf4j.version>1.7.30</slf4j.version>
         <java.level>8</java.level>
         <agent.version>2.9.5</agent.version>
         <skipTests>true</skipTests>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     </properties>
 
     <dependencies>
+
```
Comment (MarkEWaite) [inline pom.xml]: Extra blank lines are usually fine, but we generally prefer to not have trailing white space.
```suggestion

```

**Comment by MarkEWaite** [comment]
Thanks for showing me another step-wise improvement that needs to be included in the tutorial!

Many plugins depend on common libraries, like apache-httpcomponents, jackson2, aws-sdk, and snakeyaml.  They are common enough that we have Jenkins API plugins that provide those APIs as a plugin.  One of the transformations is to switch from depending on the library to instead depend on the plugin that provides the library.  It reduces the size of the plugin and keeps usage in sync across Jenkins.

File: pom.xml (hunk 2 above)
Comment (MarkEWaite) [reply pom.xml:159]: Based on the number of times that is referenced in the pom, probably want to create a new property `netty.version` and assign it the value `4.1.68.Final`, then use that property everywhere in the pom.xml file.

Comment (slaboure) [reply pom.xml:159]: So, I wasn't very happy to point to the alpha version, but AFAIR, I was forced into doing this as there is already a reference to the alpha version in the pom: 

> <dependency>
>             <groupId>io.netty</groupId>
>             <artifactId>netty-codec-http2</artifactId>
>             <version>5.0.0.Alpha2</version>
>         </dependency>

And this was conflicting with other usage of the library imported by other components

**Comment by slaboure** [comment]
> Thanks for showing me another step-wise improvement that needs to be included in the tutorial!
> 
> Many plugins depend on common libraries, like apache-httpcomponents, jackson2, aws-sdk, and snakeyaml. They are common enough that we have Jenkins API plugins that provide those APIs as a plugin. One of the transformations is to switch from depending on the library to instead depend on the plugin that provides the library. It reduces the size of the plugin and keeps usage in sync across Jenkins.

Thanks, that makes a lot of sense. Is there a list of those somewhere or is that mostly "tribal knowledge"?

**Comment by MarkEWaite** [comment]
> Thanks, that makes a lot of sense. Is there a list of those somewhere or is that mostly "tribal knowledge"?

I was just looking at the ["API" plugins on plugins.jenkins.io](https://plugins.jenkins.io/ui/search?query=api) and realized that it is "tribal knowledge".  Instructions on how to make the transition are included in several of the plugins, like [apache-httpcomponents-client-4-api](https://plugins.jenkins.io/apache-httpcomponents-client-4-api/), but there isn't a central location that lists them.  Seems like a good list to include in the tutorial section on converting API dependencies to plugin dependencies

Even that "API" query of plugins.jenkins.io is incomplete.  It doesn't include the Apache httpcomponents plugin as an API plugin.
